File: Leetcode/Python/77. Combinations/solution.py
# https://leetcode.com/problems/combinations/


from itertools import combinations

# Use itertools
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        return list(map(list, combinations(range(1, n + 1), k)))


# DFS
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        answer = []
        numbers = list(range(1, n + 1))
        def dfs(index: int, result: List[int]) -> None:
            if len(result) == k:
                answer.append(result[:])
                return
            for idx in range(index, n):
                print(idx, index)
                result.append(numbers[idx])
                dfs(idx + 1, result)
                result.pop()
        dfs(0, [])
        return answer
            

File: Leetcode/Typescript/77. Combinations/solution.ts
// https://leetcode.com/problems/combinations/description/

function combine(n: number, k: number): number[][] {
  const result: number[][] = [];

  function dfs(prev: number[], index: number) {
    if (prev.length === k) {
      result.push(prev.slice());
      return;
    }

    for (let idx = index; idx <= n; idx++) {
      prev.push(idx);
      dfs(prev, idx + 1);
      prev.pop();
    }
  }

  dfs([], 1);

  return result;
}
